### ISSUE TITLE
[@parcel/graph] SharedTypeMap should dynamically select the store data type

### ISSUE THREAD

**Issue by ahaoboy** [parcel-bundler/parcel#8621]
# 🙋 feature request
SharedTypeMap and SerializedAdjacencyList use Uint32Array to store data, but in most case Uint16Array is enough, so if can dynamically select the store data type, it's can save half memory to store and transform?
```
class SharedTypeMap {
 data: Uint32Array;
}
 
export type SerializedAdjacencyList<TEdgeType> = {|
  nodes: Uint32Array,
  edges: Uint32Array,
|};

```

## 🤔 Expected Behavior
Dynamically select the AdjacencyList's  store  data type

## 😯 Current Behavior
The AdjacencyList's  store data type is always Uint32Array, in most cases it is wasted

## 💁 Possible Solution
Each resize we can choose the appropriate data type(Uint32Array/Uint16Array) depending on the number of nodes/edges
We can hard code Uint32Array to Uint16Array and test some usecase, find if this approach is effective by comparing

## 🔦 Context
None

## 💻 Examples
I make a sample with 100_000 nodes and  edges, it looks like to be working 
<img width="1831" alt="Snipaste_2022-11-11_18-08-07" src="https://user-images.githubusercontent.com/19884146/201317620-9e7cce9f-cdd8-4783-955d-8573fe8cb562.png">

<img width="1830" alt="Snipaste_2022-11-11_18-07-29" src="https://user-images.githubusercontent.com/19884146/201317588-b76b00a4-ab0e-4a26-9ea8-55f13bb81ad5.png">

**Comment by devongovett** [comment]
@lettertwo

**Comment by lettertwo** [comment]
I like this idea! The heuristic seems simple enough. This might _just work_ for edges in particular, since they are re-hashed and copied into the new array on resize. ~~Nodes might need a bit extra work, since the 16 bit array data could not just be cloned into the 32 bit array~~ Nodes may also just work, if passing a 16 bit array to a new 32 bit array just does the right thing (it might?).

I think it has some conceptual overlap with another potential optimization, too: Edge types are always going to fit into 8 bits (or could be enforced as such), so we could save a fair amount of space if we used `DataView` instead of a fixed-size `TypedArray` (see https://github.com/parcel-bundler/parcel/pull/6922#discussion_r718652525). It would be nice to squeeze out all the extra space we can this way, but implementing on top of `DataView` brings some complications (as described in that comment).

**Comment by ahaoboy** [comment]
@lettertwo 
I think this lib can be used by more projects, but it depends some parcel utils and can't used in stackbitz, can you fix this? 
 https://github.com/stackblitz/core/issues/1855

**Comment by ahaoboy** [comment]
> I like this idea! The heuristic seems simple enough. This might _just work_ for edges in particular, since they are re-hashed and copied into the new array on resize. ~Nodes might need a bit extra work, since the 16 bit array data could not just be cloned into the 32 bit array~ Nodes may also just work, if passing a 16 bit array to a new 32 bit array just does the right thing (it might?).
> 
> I think it has some conceptual overlap with another potential optimization, too: Edge types are always going to fit into 8 bits (or could be enforced as such), so we could save a fair amount of space if we used `DataView` instead of a fixed-size `TypedArray` (see [#6922 (comment)](https://github.com/parcel-bundler/parcel/pull/6922#discussion_r718652525)). It would be nice to squeeze out all the extra space we can this way, but implementing on top of `DataView` brings some complications (as described in that comment).

Yes, in most case(may 99.99%) edeg type is less than 255, so use u8 rather than u32 can save more memory, but this adds more 
 complexity. I think it's worth a try~

**Comment by mischnic** [comment]
> I think this lib can be used by more projects, but it depends some parcel utils and can't used in stackbitz, can you fix this?

The only usage is https://github.com/parcel-bundler/parcel/blob/c3bbe0a6160186f496ca2f9e9bead9376c0522f1/packages/core/graph/src/AdjacencyList.js#L4
which imports
https://github.com/parcel-bundler/parcel/blob/c3bbe0a6160186f496ca2f9e9bead9376c0522f1/packages/core/utils/src/shared-buffer.js#L1-L24

I think we could just copy that over into the graph package.
